python/lsst/daf/butler/direct_butler.py

```diff
@@ -109,119 +109,133 @@ class ButlerValidationError(ValidationError):
-        if butler is not None:
-            if config is not None or searchPaths is not None or writeable is not None:
-                raise TypeError(
-                    "Cannot pass 'config', 'searchPaths', or 'writeable' arguments with 'butler' argument."
-                )
-            self._registry = butler._registry.copy(defaults)
-            self._datastore = butler._datastore
-            self.storageClasses = butler.storageClasses
-            self._config: ButlerConfig = butler._config
-        else:
-            self._config = ButlerConfig(config, searchPaths=searchPaths, without_datastore=without_datastore)
-            try:
-                butlerRoot = self._config.get("root", self._config.configDir)
-                if writeable is None:
-                    writeable = run is not None
-                self._registry = _RegistryFactory(self._config).from_config(
-                    butlerRoot=butlerRoot, writeable=writeable, defaults=defaults
-                )
-                if without_datastore:
-                    self._datastore = NullDatastore(None, None)
-                else:
-                    self._datastore = Datastore.fromConfig(
-                        self._config, self._registry.getDatastoreBridgeManager(), butlerRoot=butlerRoot
-                    )
-                # TODO: Once datastore drops dependency on registry we can
-                # construct datastore first and pass opaque tables to registry
-                # constructor.
-                self._registry.make_datastore_tables(self._datastore.get_opaque_table_definitions())
-                self.storageClasses = StorageClassFactory()
-                self.storageClasses.addFromConfig(self._config)
-            except Exception:
-                # Failures here usually mean that configuration is incomplete,
-                # just issue an error message which includes config file URI.
-                _LOG.error(f"Failed to instantiate Butler from config {self._config.configFile}.")
-                raise
+        config: ButlerConfig,
+        registry: SqlRegistry,
+        datastore: Datastore,
+        storageClasses: StorageClassFactory,
+    ) -> DirectButler:
+        self = cast(DirectButler, super().__new__(cls))
+        self._config = config
+        self._registry = registry
+        self._datastore = datastore
+        self.storageClasses = storageClasses
 
         # For execution butler the datastore needs a special
         # dependency-inversion trick. This is not used by regular butler,
         # but we do not have a way to distinguish regular butler from execution
         # butler.
         self._datastore.set_retrieve_dataset_type_method(self._retrieve_dataset_type)
 
-        if "run" in self._config or "collection" in self._config:
+        self._registry_shim = RegistryShim(self)
+
+        return self
+
+    @classmethod
+    def create_from_config(
```

Comment:
I wanted to name this `from_config` but mypy wouldn't let me because it has a different signature than `Butler.from_config`.  There is no public `create_from_config` -- it's private because it's only on `DirectButler`.

Reply:
It is public in the sense that:

```python
>>> b = Butler("dp02")
>>> help(b)
```
will show the method, and presumably Jupyter Notebooks will report the method. Maybe add an extra note to the docstring pointing to `Butler.from_config` as the general interface.